File: sdks/java/src/main/java/ai/beater/sdk/Beater.java
package ai.beater.sdk;

import io.opentelemetry.api.OpenTelemetry;
import io.opentelemetry.api.trace.Span;
import io.opentelemetry.api.trace.StatusCode;
import io.opentelemetry.api.trace.Tracer;
import io.opentelemetry.context.Scope;
import io.opentelemetry.exporter.otlp.http.trace.OtlpHttpSpanExporter;
import io.opentelemetry.sdk.OpenTelemetrySdk;
import io.opentelemetry.sdk.resources.Resource;
import io.opentelemetry.sdk.trace.SdkTracerProvider;
import io.opentelemetry.sdk.trace.SpanProcessor;
import io.opentelemetry.sdk.trace.export.BatchSpanProcessor;
import java.util.concurrent.atomic.AtomicLong;
import java.util.function.Supplier;

/**
 * Ergonomic entry point for tracing arbitrary code as Beater spans.
 *
 * <p>Call {@link #init(BeaterConfig)} once at process start, then wrap work in
 * {@link #observe(String, String, Supplier)}.
 */
public final class Beater {

    private static final String TRACER_NAME = "beater.sdk";

    private static volatile OpenTelemetrySdk sdk;
    private static volatile SdkTracerProvider provider;
    private static volatile BeaterConfig config;
    private static volatile Tracer tracer;
    private static final AtomicLong SEQ = new AtomicLong(0);

    private Beater() {}

    /** Initialize the Beater tracer with an OTLP/HTTP exporter. Call once at startup. */
    public static synchronized void init(BeaterConfig cfg) {
        var exporterBuilder = OtlpHttpSpanExporter.builder().setEndpoint(cfg.otlpHttpTracesUrl());
        if (cfg.apiKey != null && !cfg.apiKey.isEmpty()) {
            exporterBuilder.addHeader("authorization", "Bearer " + cfg.apiKey);
        }
        SpanProcessor processor = BatchSpanProcessor.builder(exporterBuilder.build()).build();
        configure(cfg, processor);
    }

    /**
     * Initialize with a caller-supplied {@link SpanProcessor}. Useful for tests
     * (e.g. wiring an in-memory exporter) and custom export pipelines.
     */
    public static synchronized void init(BeaterConfig cfg, SpanProcessor processor) {
        configure(cfg, processor);
    }

    private static void configure(BeaterConfig cfg, SpanProcessor processor) {
        Resource resource = Resource.getDefault().toBuilder()
                .put("service.name", cfg.serviceName)
                .build();
        SdkTracerProvider tracerProvider = SdkTracerProvider.builder()
                .setResource(resource)
                .addSpanProcessor(processor)
                .build();
        OpenTelemetrySdk built = OpenTelemetrySdk.builder()
                .setTracerProvider(tracerProvider)
                .build();
        provider = tracerProvider;
        sdk = built;
        config = cfg;
        tracer = built.getTracer(TRACER_NAME);
    }

    private static Tracer tracer() {
        if (tracer == null) {
            init(BeaterConfig.fromEnv());
        }
        return tracer;
    }

    private static void applyCommon(Span span, String kind) {
        span.setAttribute(SemConv.SPAN_KIND, kind);
        span.setAttribute(SemConv.SEQ, SEQ.incrementAndGet());
        BeaterConfig cfg = config;
        if (cfg != null && cfg.releaseId != null && !cfg.releaseId.isEmpty()) {
            span.setAttribute(SemConv.RELEASE_ID, cfg.releaseId);
        }
    }

    /** Open a span, run {@code body}, record status, and end the span; returns the result. */
    public static <T> T observe(String name, String kind, Supplier<T> body) {
        Span span = tracer().spanBuilder(name).startSpan();
        applyCommon(span, kind);
        try (Scope ignored = span.makeCurrent()) {
            T result = body.get();
            span.setStatus(StatusCode.OK);
            return result;
        } catch (RuntimeException e) {
            span.setStatus(StatusCode.ERROR, String.valueOf(e.getMessage()));
            span.recordException(e);
            throw e;
        } finally {
            span.end();
        }
    }

    /** {@link Runnable} overload of {@link #observe(String, String, Supplier)}. */
    public static void observe(String name, String kind, Runnable body) {
        observe(name, kind, () -> {
            body.run();
            return null;
        });
    }

    /** Attach an input payload to the current span (JSON-encoded if not a String). */
    public static void setInput(Object value) {
        Span.current().setAttribute(SemConv.INPUT_VALUE, toValue(value));
    }

    /** Attach an output payload to the current span (JSON-encoded if not a String). */
    public static void setOutput(Object value) {
        Span.current().setAttribute(SemConv.OUTPUT_VALUE, toValue(value));
    }

    /** Force-flush pending spans. Useful before a short-lived process exits. */
    public static void flush() {
        SdkTracerProvider p = provider;
        if (p != null) {
            p.forceFlush().join(30, java.util.concurrent.TimeUnit.SECONDS);
        }
    }

    /** Shut down the tracer pipeline, flushing any pending spans. */
    public static synchronized void shutdown() {
        OpenTelemetrySdk s = sdk;
        if (s != null) {
            s.close();
        }
        sdk = null;
        provider = null;
        tracer = null;
        config = null;
    }

    /** String values pass through; everything else is encoded as a small JSON value. */
    static String toValue(Object obj) {
        if (obj == null) {
            return "null";
        }
        if (obj instanceof String s) {
            return s;
        }
        if (obj instanceof Number || obj instanceof Boolean) {
            return obj.toString();
        }
        return jsonString(String.valueOf(obj));
    }

    private static String jsonString(String s) {
        StringBuilder sb = new StringBuilder(s.length() + 2);
        sb.append('"');
        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            switch (c) {
                case '"' -> sb.append("\\\"");
                case '\\' -> sb.append("\\\\");
                case '\n' -> sb.append("\\n");
                case '\r' -> sb.append("\\r");
                case '\t' -> sb.append("\\t");
                default -> {
                    if (c < 0x20) {
                        sb.append(String.format("\\u%04x", (int) c));
                    } else {
                        sb.append(c);
                    }
                }
            }
        }
        sb.append('"');
        return sb.toString();
    }
}

File: sdks/go/beater.go
package beater

import (
	"context"
	"encoding/json"
	"strings"
	"sync/atomic"

	"go.opentelemetry.io/otel"
	"go.opentelemetry.io/otel/attribute"
	"go.opentelemetry.io/otel/codes"
	"go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracehttp"
	"go.opentelemetry.io/otel/sdk/resource"
	sdktrace "go.opentelemetry.io/otel/sdk/trace"
	semconv "go.opentelemetry.io/otel/semconv/v1.26.0"
	"go.opentelemetry.io/otel/trace"
)

const tracerName = "beater.sdk"

// Package-level state set by Init. The provider is also stashed so Flush can
// force-flush without the caller holding the shutdown closure.
var (
	cfg      atomic.Pointer[Config]
	provider atomic.Pointer[sdktrace.TracerProvider]
	seq      atomic.Int64
)

// Init configures a global OpenTelemetry TracerProvider that exports spans to
// Beater over OTLP/HTTP. It returns a shutdown function that flushes and stops
// the provider; call it (e.g. with defer) before the process exits.
func Init(ctx context.Context, c Config) (func(context.Context) error, error) {
	opts := []otlptracehttp.Option{
		otlptracehttp.WithEndpointURL(c.otlpTracesURL()),
	}
	if strings.HasPrefix(c.BaseURL, "http://") {
		opts = append(opts, otlptracehttp.WithInsecure())
	}
	if c.APIKey != "" {
		opts = append(opts, otlptracehttp.WithHeaders(map[string]string{
			"authorization": "Bearer " + c.APIKey,
		}))
	}

	exporter, err := otlptracehttp.New(ctx, opts...)
	if err != nil {
		return nil, err
	}

	res, err := resource.New(ctx, resource.WithAttributes(
		semconv.ServiceName(c.ServiceName),
	))
	if err != nil {
		return nil, err
	}

	tp := sdktrace.NewTracerProvider(
		sdktrace.WithBatcher(exporter),
		sdktrace.WithResource(res),
	)
	otel.SetTracerProvider(tp)

	conf := c
	cfg.Store(&conf)
	provider.Store(tp)

	return func(ctx context.Context) error {
		return tp.Shutdown(ctx)
	}, nil
}

func tracer() trace.Tracer {
	return otel.GetTracerProvider().Tracer(tracerName)
}

func applyCommon(span trace.Span, kind string) {
	span.SetAttributes(
		attribute.String(AttrSpanKind, kind),
		attribute.Int64(AttrSeq, seq.Add(1)),
	)
	if c := cfg.Load(); c != nil && c.ReleaseID != "" {
		span.SetAttributes(attribute.String(AttrReleaseID, c.ReleaseID))
	}
}

func toValue(v any) string {
	if s, ok := v.(string); ok {
		return s
	}
	b, err := json.Marshal(v)
	if err != nil {
		return ""
	}
	return string(b)
}

// Observe runs fn inside a new span named name with the given span kind. It sets
// the kind/seq/release attributes, records the error and span status, and ends
// the span. The error returned by fn is propagated.
func Observe(ctx context.Context, name, kind string, fn func(ctx context.Context) error) error {
	ctx, span := tracer().Start(ctx, name)
	defer span.End()
	applyCommon(span, kind)

	if err := fn(ctx); err != nil {
		span.RecordError(err)
		span.SetStatus(codes.Error, err.Error())
		return err
	}
	span.SetStatus(codes.Ok, "")
	return nil
}

// SetInput attaches an input payload to the span in ctx. Non-string values are
// JSON-encoded.
func SetInput(ctx context.Context, v any) {
	trace.SpanFromContext(ctx).SetAttributes(attribute.String(AttrInputValue, toValue(v)))
}

// SetOutput attaches an output payload to the span in ctx. Non-string values are
// JSON-encoded.
func SetOutput(ctx context.Context, v any) {
	trace.SpanFromContext(ctx).SetAttributes(attribute.String(AttrOutputValue, toValue(v)))
}

// Flush force-flushes any buffered spans to the exporter. Useful before a
// short-lived program exits.
func Flush(ctx context.Context) error {
	if tp := provider.Load(); tp != nil {
		return tp.ForceFlush(ctx)
	}
	return nil
}
